contracts/JBV3Token.sol

```diff
@@ -0,0 +1,401 @@
+
+  /** 
+    @notice
+    v2ProjectId to migrate from. 
+    v1ProjectId to migrate from. 
+  */
+  struct MigrationInfo {
+    uint128 v2ProjectId;
+    uint128 v1ProjectId;
+  }
+
+  //*********************************************************************//
+  // ------------------------- external views -------------------------- //
+  //*********************************************************************//
+
+  /** 
+    @notice
+    The total supply of this ERC20.
+
+    @param _projectId the ID of the project to which the token belongs. This is ignored.
+
+    @return The total supply of this ERC20, as a fixed point number.
+  */
+  function totalSupply(uint256 _projectId) external view override returns (uint256) {
+    _projectId; // Prevents unused var compiler and natspec complaints.
+
+    (uint128 v2ProjectId, uint128 v1ProjectId) = getMigrationInfo(projectId);
+
+    return super.totalSupply() + 
+    v1TicketBooth.totalSupplyOf(v1ProjectId) + 
+    v2TokenStore.totalSupplyOf(v2ProjectId) -
+    v1TicketBooth.balanceOf(address(this), projectId) -
+    v2TokenStore.balanceOf(address(this), projectId);
+  }
+
+  /** 
+    @notice
+    The total supply of this ERC20.
+
+    @return The total supply of this ERC20, as a fixed point number.
+  */
+  function totalSupply() public view override returns (uint256) {
+    (uint128 v2ProjectId, uint128 v1ProjectId) = getMigrationInfo(projectId);
+
+    return super.totalSupply() + 
+    v1TicketBooth.totalSupplyOf(v1ProjectId) + 
+    v2TokenStore.totalSupplyOf(v2ProjectId) -
+    v1TicketBooth.balanceOf(address(this), projectId) -
+    v2TokenStore.balanceOf(address(this), projectId);
+  }
+
+  /** 
+    @notice
+    An account's balance of this ERC20.
+
+    @param _account The account to get a balance of.
+    @param _projectId is the ID of the project to which the token belongs. This is ignored.
+
+    @return The balance of the `_account` of this ERC20, as a fixed point number with 18 decimals.
+  */
+  function balanceOf(address _account, uint256 _projectId)
+    external
+    view
+    override
+    returns (uint256)
+  {
+    _projectId; // Prevents unused var compiler and natspec complaints.
+
+    return super.balanceOf(_account);
+  }
+
+  //*********************************************************************//
+  // -------------------------- public views --------------------------- //
+  //*********************************************************************//
+
+  /** 
+    @notice
+    The number of decimals included in the fixed point accounting of this token.
+
+    @return The number of decimals.
+  */
+  function decimals() public view override(ERC20, IJBToken) returns (uint8) {
+    return super.decimals();
+  }
+
+  /** 
+    @notice
+    The token id's to migrate from.
+
+    @return v1 & v2 project id's.
+  */
+  function getMigrationInfo(uint256 _projectId) public view returns(uint128, uint128) {
+    MigrationInfo memory _migrationInfo = migrationOf[_projectId];
```

Comment:
cheaper to not copy the struct in memory -> either:
`return (migrationOf[_projectId].v2ProjectId, migrationOf[_projectId].v1ProjectId);` (cost 24397 or 23709 with 200runs)
or use the storage ref:
```
MigrationInfo storage _migrationInfo = migrationOf[_projectId];
return (_migrationInfo.v2ProjectId, _migrationInfo.v1ProjectId);
 ```
 (cost 24332 - 23709)

Reply:
current one costs 24435 (23772 with 200runs optim)